lib/sqlalchemy/ext/orderinglist.py

```diff
@@ -345,48 +351,47 @@ def _raw_append(self, entity):
 
     _raw_append = collection.adds(1)(_raw_append)
 
-    def insert(self, index, entity):
-        super().insert(index, entity)
+    def insert(self, __index: SupportsIndex, __entity: _T) -> None:
+        super().insert(__index, __entity)
         self._reorder()
 
-    def remove(self, entity):
-        super().remove(entity)
+    def remove(self, __entity: _T) -> None:
+        super().remove(__entity)
 
         adapter = collection_adapter(self)
         if adapter and adapter._referenced_by_owner:
             self._reorder()
 
-    def pop(self, index=-1):
-        entity = super().pop(index)
+    def pop(self, __index: SupportsIndex = -1) -> _T:
+        entity = super().pop(__index)
         self._reorder()
         return entity
 
-    def __setitem__(self, index, entity):
-        if isinstance(index, slice):
-            step = index.step or 1
-            start = index.start or 0
-            if start < 0:
-                start += len(self)
-            stop = index.stop or len(self)
-            if stop < 0:
-                stop += len(self)
-
-            for i in range(start, stop, step):
-                self.__setitem__(i, entity[i])
-        else:
-            self._order_entity(index, entity, True)
-            super().__setitem__(index, entity)
+    @overload
+    def __setitem__(self, __index: SupportsIndex, __entity: _T) -> None: ...
 
-    def __delitem__(self, index):
-        super().__delitem__(index)
-        self._reorder()
+    @overload
+    def __setitem__(self, __index: slice, __entity: Iterable[_T]) -> None: ...
 
-    def __setslice__(self, start, end, values):
-        super().__setslice__(start, end, values)
-        self._reorder()
+    def __setitem__(
+        self,
+        __index: Union[SupportsIndex, slice],
+        __entity: Union[_T, Iterable[_T]],
+    ) -> None:
+        if isinstance(__index, slice):
+            # there are enough edge cases with slice lengths longer or shorter
+            # than the length of the assigned items, that full-on reordering
+            # is far simpler.
+            # SQLAlchemy collection instrumentation otherwise takes care of
+            # slices in all normal use of OrderingList in ORM classes.
+            super().__setitem__(__index, __entity)
+            self._reorder()
+        else:
```

Comment:
**Martijn Pieters** (mjpieters) wrote:

We'd have to test against an _uninstrumented_, raw `OrderingList` instance. That's because when used in the context of a ORM relationship the `.orm.collections._list_decorators()` function provides a `__setitem__` implementation that maps a `slice` instance to `self.__delitem__(..)` / `self.insert(..)` or `self.__setitem__(..)` calls for each affected index in the slice, and so the slice object never reaches this point.

It's been a while since I touched this PR, so I can't quite recall the order of events, but I think I hadn't yet realised that relationship instrumentation meant that the issues with slicing are not seen in regular use when I created the ticket, and I only hit upon the ORM relationship instrumentation when I was writing the PR and saw there were tests involving slices.

I'll have to find some time to create such a test.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5118